Userland/Services/AudioServer/Mixer.cpp

```diff
@@ -65,13 +66,12 @@ void Mixer::mix()
     decltype(m_pending_mixing) active_mix_queues;
 
     for (;;) {
-        if (active_mix_queues.is_empty() || m_added_queue) {
-            pthread_mutex_lock(&m_pending_mutex);
-            pthread_cond_wait(&m_pending_cond, &m_pending_mutex);
+        m_pending_mutex.lock();
+        // While we have nothing to mix, wait on the condition.
+        m_mixing_necessary.wait_while([this, &active_mix_queues]() { return m_pending_mixing.is_empty() && active_mix_queues.is_empty(); });
```

Comment:
I'm not too familiar with AudioServer internals, but can `active_mix_queues` possibly _not_ be empty here? Wouldn't we have handled that in the previous iteration?

And if I'm right about that, then the check below is redundant.

But other than that, yes, this looks correct 😀

Reply:
The reason we pause on active_mix_queues being empty is the whole point of this condition variable. The mixer thread will wait until a client connects, which will make the audio server not run while no audio has to be mixed and streamed. The check below is so that we only add a new queue if necessary. (There's now no condition around the entire lock thing, so we need to prevent unnecessary busywork). The condition will e.g. return immediately if we're mixing audio like normal, i.e. active_mix_queues has 1 or 2 clients and no client was added.

Reply:
Right, I see. The logic below doesn't drain the `active_mix_queues` list, it drains the queues themselves. So what you're doing here is if you already have active clients, you do a quick non-blocking check for new clients, and otherwise you wait. Good.

What would happen if you have some queues, but they are all empty? Is that possible?

Reply:
If a queue is empty, we don't append audio data for it (line 97). If that happens with all clients, we just don't change the current mixing buffer and stream all zeroes to the sound card. It's definitely possible, but I don't even know if it can happen with SoundPlayer while no file is loaded. That's a further optimization we could do, but it doesn't do harm because the few programs that do actually connect to the audio server usually want to stream audio.